.github/workflows/main.yml

```diff
@@ -151,11 +155,85 @@ jobs:
           python: 'pypy3.9'
           allow_failure: false
 
+  report-coverage:
+    name: Report Combined Coverage
+    runs-on: ubuntu-24.04
+    needs: test
+    if: always()
+    permissions:
+      contents: read
+      actions: write
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.13'
+
+      - name: Install coverage tool
+        run: python -m pip install coverage[toml]
+
+      - name: Download all coverage artifacts
+        uses: actions/download-artifact@v4
+        with:
+          path: downloaded-coverage-artifacts
+          pattern: coverage-artifact-*
+
+      - name: Combine coverage reports
+        run: |
+          mkdir combined_coverage_data
+          find downloaded-coverage-artifacts -type f -name 'coverage.*' -exec cp {} combined_coverage_data/ \;
+          coverage combine combined_coverage_data/*
+          coverage xml
+          coverage html
+          coverage report --show-missing --format=markdown >> $GITHUB_STEP_SUMMARY
+
+      - name: Determine retention days
+        id: determine-retention-days
+        run: |
+          if [ "${GITHUB_REF}" = "refs/heads/main" ] || [[ "${GITHUB_REF}" == refs/tags/* ]]; then
+            echo "retention_days=90" >> $GITHUB_OUTPUT
+          else
+            echo "retention_days=3" >> $GITHUB_OUTPUT
+          fi
+        env:
+          GITHUB_REF: ${{ github.ref }}
+
+      - name: Upload combined .coverage file
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-file
+          path: .coverage
+          retention-days: ${{ steps.determine-retention-days.outputs.retention_days }}
+          include-hidden-files: true
+
+      - name: Upload HTML coverage report
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-html-report
+          path: htmlcov
+          retention-days: ${{ steps.determine-retention-days.outputs.retention_days }}
+
+      - name: Delete temporary coverage artifacts from run
```

Comment:
This isn't really that necessary, as you already set them to expire in one day. Plus, it'd be impossible to inspect these files whenever you suspect something's wrong.

Reply:
True. But its just noise, i figure i'd leave it "clean" -- but you're not wrong.

Reply:
It's just that I've recently seen a DIY implementation of this causing CI stability issues due to GH API or network being flaky and causing chaos. That memory is still fresh :)

But also, these are about inspectability.

If you want fewer artifacts, you can run `actions/upload-artifact/merge` and it'll do the deletion for you, replacing these with one artifact containing all the individual files.

Reply:
Now that i like! Thanks!